File: 백준/Gold/6198. 옥상 정원 꾸미기/옥상 정원 꾸미기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        long[] heights = new long[n];
        for (int i = 0; i < n; i++) {
            heights[i] = Long.parseLong(br.readLine());
        }

        Stack<Long> stack = new Stack<>();
        long totalCount = 0;

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && stack.peek() <= heights[i]) {
                stack.pop();
            }

            totalCount += stack.size();

            stack.push(heights[i]);
        }

        System.out.print(totalCount);
    }
}

File: 백준/Gold/6198. 옥상 정원 꾸미기/옥상 정원 꾸미기.kt
import java.util.*

private var n = 0
private lateinit var stack: Stack<Long>

fun main() {
    val br = System.`in`.bufferedReader()
    n = br.readLine().toInt()
    stack = Stack()
    stack.push(Long.MAX_VALUE)
    var cnt :Long = 0
    repeat(n){
        val cur = br.readLine().toLong()
        while(stack.peek() <= cur)
            stack.pop()
        cnt += (stack.size -1)
        stack.push(cur)
    }
    println(cnt)
}